src/wallet/spend.cpp

```diff
@@ -855,6 +855,13 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
 
     coin_selection_params.m_min_change_target = GenerateChangeTarget(std::floor(recipients_sum / vecSend.size()), coin_selection_params.m_change_fee, rng_fast);
 
+    // The smallest change amount should be:
+    // 1. at least equal to dust threshold
+    // 2. at least 1 sat greater than fees to spend it at m_discard_feerate
+    const auto dust = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
+    const auto change_spend_fee = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size);
+    coin_selection_params.min_viable_change = std::max(change_spend_fee + 1, dust);
```

Comment:
hmm, this doesn't seems to be right:

1) The dust amount is taken from --> `GetDustThreshold` which receives the change output, serializes it and uses magic numbers to calculate the input vsize (it's a general minimum input vsize calculation for an output), then calls `discard_fee.Get(vsize)` with it to get the dust amount.

2. The `change_spend_fee` is taken from --> `discard_fee.Get(change_spend_size)` , where `change_spend_size` is the result of `CalculateMaximumSignedInputSize(output)` which is the real vsize calculation of a signed input that spends the change output (we don't hardcode numbers here, we create the input and sign it).


In other words:

The only difference that I'm seeing between the `dust` and `change_spend_fee` variables is the vsize calculation (both are doing `m_discard_fee.GetFee(vsize)`), and:

- for `dust`, vsize is calculated from hardcoded numbers to get the minimum theoretical possible size.
- for `change_spend_fee`, vsize is calculated creating an input and dummy signing it.

So, `dust` will always be smaller than `change_spend_fee` right?

Reply:
Extra note, discussed this topic with @theStack for a while. He mentioned that you might wanted to use `wallet.chain().relayDustFee()` instead of `coin_selection_params.m_discard_feerate` in the dust amount calculation?

Reply:
What's implemented here matches the behavior in master. For this PR, I think this is correct.

However we should probably use `m_long_term_feerate` for calculating the change spend fee rather than the discard feerate. This would make it match how we calculate "future spends" for all of our inputs.

> you might wanted to use `wallet.chain().relayDustFee()` instead of `coin_selection_params.m_discard_feerate` in the dust amount calculation?

I think discard feerate is correct here. It is the maximum of the longest term fee estimate and the dust relay fee. As the name suggests, it is the feerate that we are willing to discard at, so I think it makes sense to keep it as is.

Reply:
> > you might wanted to use `wallet.chain().relayDustFee()` instead of `coin_selection_params.m_discard_feerate` in the dust amount calculation?
> 
> I think discard feerate is correct here. It is the maximum of the longest term fee estimate and the dust relay fee. As the name suggests, it is the feerate that we are willing to discard at, so I think it makes sense to keep it as is.

Isn't the point of setting the smallest change amount to "at least equal to dust threshold"  to ensure that the created tx can enter our own mempool now, rather than somewhen in the future (considering that we run a node with a possibly customized `-dustrelayfee=...` option)? As an example, if one would set `-dustrelayfee=...` to a high value and `discardfee=0`, a tx with a change output could be created that is rejected from our mempool (due to being `IsDust` and therefore `!IsStandardTx()`) and thus can't be sent. That's why my assumption was that we want to call `GetDustThreshold` with the dustRelayFee instead:

```diff
-    const auto dust = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
+    const auto dust = GetDustThreshold(change_prototype_txout, wallet.chain.relayDustFee());
```
(Am I missing something?)

Reply:
> As an example, if one would set `-dustrelayfee=...` to a high value and `discardfee=0`, a tx with a change output could be created that is rejected from our mempool

Then the dustrelayfee is what `m_discard_fee` is set to. See `GetDiscardRate` in src/wallet/fees.cpp.

Reply:
> > As an example, if one would set `-dustrelayfee=...` to a high value and `discardfee=0`, a tx with a change output could be created that is rejected from our mempool
> 
> Then the dustrelayfee is what `m_discard_fee` is set to. See `GetDiscardRate` in src/wallet/fees.cpp.

Oh right, I completely overlooked this following line of code, d'oh! 🤦‍♂️ 

https://github.com/bitcoin/bitcoin/blob/a8f69541ad53a76d5f69044ba829069d225a4af1/src/wallet/fees.cpp#L91

Reply:
I believe we should get rid of `discard fee rate` and replace it with `dust relay fee rate` and `long term fee rate`. For this PR I decide to stick to the current implementation to limit the scope. I'll open an issue to deprecate `discard fee rate` if there is none yet.